Add section on network learning

stefpeschel opened the pull request:
Hey,
As announced, Christian Müller (@muellsen) and I would like to contribute a section on network learning to OMA. 
Let me know if I should modify anything.

As I wrote on Gitter, the approach involves permutation testing, which takes several hours to run. Also, each network construction takes about a minute. So these objects are stored in a new folder "general/network_files" and loaded when the book is knitted. I hope this is okay.

Best,
Stefanie

antagomir commented:
Many thanks. The solution seems very reasonable for now, we can go with this until someone comes up with better options. Let us review the contents next..

antagomir commented on book.bib:
```diff
@@ -1397,3 +1397,49 @@ @Article{ZIBB2018
 
+
+@article{yoon2019microbial,
+  title={Microbial networks in SPRING-Semi-parametric rank-based correlation and partial correlation estimation for quantitative microbiome data},
```
Should the title be in double curly brackets so that the capitalizations will show correctly? {{}}

antagomir commented on book.bib:
```diff
@@ -1397,3 +1397,49 @@ @Article{ZIBB2018
+@article{yoon2019microbial,
+  title={Microbial networks in SPRING-Semi-parametric rank-based correlation and partial correlation estimation for quantitative microbiome data},
+  author={Yoon, Grace and Gaynanova, Irina and M{\"u}ller, Christian L},
+  journal={Frontiers in genetics},
```
genetics -> Genetics?

antagomir commented on book.bib:
```diff
@@ -1397,3 +1397,49 @@ @Article{ZIBB2018
+
+@article{peschel2021netcomi,
+    author = {Peschel, Stefanie and M{\"{u}}ller, Christian L and von Mutius, Erika and Boulesteix, Anne-Laure and Depner, Martin},
+    title = "{NetCoMi: network construction and comparison for microbiome data in R}",
```
Same, check double curly brackets since capitalization carries meaning here. 

The quotes are unnecessary I think?

antagomir replied:
Same, check double curly brackets since capitalization carries meaning here. 

The quotes are unnecessary I think?

antagomir commented on 60_networks.Rmd:
```diff
@@ -0,0 +1,664 @@
+
+| Function               | Purpose                | Input
+|:----------------------:|:----------------------:|:----------------------:|
+| `netConstruct`       | Network construction   | Count matrix, phyloseq object, or association matrix|
```
This refers to phyloseq but the OMA contents are not based on _phyloseq_ but instead of the alternative data container, _TreeSummarizedExperiment_. Hence this line may be confusing for new users who are not familiar with the differences. Does the `netConstruct` function support _TreeSummarizedExperiment_, or could this be added? We can provide help with this and it should be relatively straightforward to add support for TreeSE containers, too.

antagomir commented on 60_networks.Rmd:
````diff
@@ -0,0 +1,664 @@
+```{r load_packages, message=FALSE, warning=FALSE}
+library(NetCoMi)
+library(mia)
+library(phyloseq)
````
We do support translations between these two containers but this book does not support phyloseq, or phyloseq-based tools in general because this would lead to mixing of two rather different tool 
 ecosystems (TreeSE vs phyloseq) would be confusing from user perspective.

antagomir commented on 60_networks.Rmd:
````diff
@@ -0,0 +1,664 @@
+data("peerj13075", package = "mia")
+```
+
+Since NetCoMi expects either a count matrix or a phyloseq object as input, we first transform the data into a phyloseq object. This object contains the count matrix, the taxonomy table, and a list with metadata.
+
+```{r make_phyloseq}
+peerj_phy <- makePhyloseqFromTreeSE(peerj13075)
+peerj_phy
+head(tax_table(peerj_phy))
+```
+Since we want to construct the networks at genus level, we need to make the genus names unique. NetCoMi's function `renameTaxa` does this job for us by numerating duplicate names. 
+
````
Ok - let us discuss this over email, and see how to solve the phyloseq issue.

antagomir commented on 60_networks.Rmd:
```diff
@@ -0,0 +1,664 @@
+* Only samples with a total of at least 1000 reads are included (argument `filtSamp`).
+* Only the 50 most abundant taxa are included (argument `filtTax`).
+
+The argument `measure` defines the association or dissimilarity measure, in our case the SPRING approach. Additional arguments are passed to `SPRING` via `measurePar`. `nlambda` and `rep.num` are the length of the $\lambda$ vector and the repetition number of subsampling used for model selection.The "approx" method (multilinear approximation of latent correlations) is used to save runtime.
+
+Normalization and zero handling is done internally in `SPRING`, so we set `normMethod` and `zeroMethod` to "none".
+
+We also set `sparsMethod` to "none" because the SPRING approach includes model selection to obtain a sparse network. Therefore, no additional sparsification step is needed.
+
+The "signed" method is used to transform associations into dissimilarities (argument `dissFunc`). Thus, strongly negatively associated taxa have a high dissimilarity and conversely a low similarity, corresponding to low edge weights on the network plot.
+
```
Consider carefully how much of these technical details to provide here vs. the package documentation. 

The general philosophy:
1. OMA provides general guidance on implementing useful analyses and often this focuses on the application problem & general concepts more than individual package. OMA links to the individual packages for more information.
2. The R packages own website/vignette/manpages will contain more technical details and examples that complement and expand material in OMA.

Here, I would suggest moving some of the details to package's own tutorial and linking from here. There are no exact guidelines, let us consider it from the user perspective and view OMA as a general, approachable introduction to microbiome data science methodology.

antagomir commented on 60_networks.Rmd:
```diff
@@ -0,0 +1,664 @@
+The `microNet` object is now passed to `netAnalyze` to perform network analysis.
+
+The function computes several common network analysis measures such as centrality measures, cluster assignment, the graphlet correlation matrix, as well as global network measures like the average path length and the clustering coefficient.
+
+Here, `centrLCC` is set to `TRUE` meaning that centralities are calculated only 
+for nodes in the largest connected component (LCC). 
+
+Clusters are identified using greedy modularity optimization 
+(by `cluster_fast_greedy` from [`igraph`](https://igraph.org/r/) package).
+
+Hubs are nodes with an eigenvector centrality value above the empirical 
+95% quantile of all eigenvector centralities in the network (argument `hubPar`).
+
+`weightDeg` and `normDeg` are set to `FALSE` so that the degree of a node is
+simply the number of its adjacent nodes.
+
+By default, a heatmap of the Graphlet Correlation Matrix (GCM) is returned (with 
+graphlet correlations in the upper triangle and significance codes resulting 
+from Student's t-test in the lower triangle). 
+See `?calcGCM` and `?testGCM` for details.
```
Same with these options - avoid replicating argument details that are easily found from the package documentation itself. Instead of explaining each argument, consider showing some illustrative complementary examples on what is possible to do with these different options.

antagomir commented on 60_networks.Rmd:
```diff
@@ -0,0 +1,664 @@
+
+Interpretation of some of the results:
+
+* The largest connected component (LCC) has 35 nodes and the network contains 6 singletons.
+* 10 clusters have been identified, containing two to eight nodes.
+* The detected hub nodes are Citrobacter, Escherichia, and Salmonella, which by definition are the nodes with the highest eigenvector centrality.
+* These three genera do also have a high degree and closeness centrality, but a lower betweenness centrality.
+* The average path length is 3.56. This means that on average it takes 3.56 steps (step length is the average dissimilarity) to get from one node to another node in the network.
```
In case these numbers may change with random seed, consider automating the numbers in the text?

antagomir replied:
Like instead of "35 nodes", it would be "`r nnodes` nodes" and the numbers automatically rendered

antagomir commented on 60_networks.Rmd:
```diff
@@ -0,0 +1,664 @@
+
+In the first plot, node colors and sizes are used to visualize certain network properties.
+
+Here is a short explanation of our argument configuration:
+
+* `layout = "spring"`: A force-directed layout is used, placing strongly connected nodes close together and nodes with less edge weight farther apart.
+* `repulsion = 0.98`: Default is 1. By decreasing the repulsion, nodes are placed farther apart.
+* `rmSingles = TRUE`: Singletons are removed.
+* `labelScale = FALSE`: Labels are not scaled by node size.
+* `nodeSize = "eigenvector"`: Nodes are scaled by eigenvector centrality.
+* `nodeSizeSpread = 3`: Node sizes are more similar if the value is decreased. This argument (in combination with cexNodes) is useful to enlarge small nodes while keeping the size of big nodes. Default is 4.
+* `nodeColor = "cluster"`: Node colors represent clusters.
+* `edgeTranspHigh = 20`: Edges with high edge weight are given some transparency so that node labels overlapping the edges are still readable.
+* `cex...`: Scaling of the node sizes, labels, title, etc.
+
+**Note:** Nodes are sometimes placed too close together so that the labels overlap. You may need to play around with the repulsion until you find a value where the labels are legible, but also the clusters are still well recognizable.
```
Again, consider citing function documentation in the package instead of replicating argument descriptions here.

antagomir commented on 60_networks.Rmd:
```diff
@@ -0,0 +1,664 @@
+     cexHubLabels = 1.3,
+     mar = c(1, 3, 4, 8))
+
+legend(0.7, 1.1, cex = 1.7, title = "estimated correlation:",
```
prefer capitalizing the title

antagomir commented on 60_networks.Rmd:
```diff
@@ -0,0 +1,664 @@
+
+#### Highlight data features
+
+We now color nodes according to their phylum. The sizes of the nodes are scaled by the mclr-transformed abundances of the genera, since SPRING uses this transformation to account for compositionality.
```
mclr?

antagomir replied:
Usually clr or rclr are used, it would be helpful to explain here what mclr is

antagomir commented on 60_networks.Rmd:
```diff
@@ -0,0 +1,664 @@
+# Define phylum colors
+phylcol <- c("cyan", "blue3", "red", "lawngreen", "yellow", "deeppink", "orange")
+
+plot(peerj_netprops,
+     layout = "spring",
+     repulsion = 0.98,
+     rmSingles = TRUE,
+     labelScale = FALSE,
+     nodeSize = "mclr",
+     nodeColor = "feature", 
+     featVecCol = phyla, 
+     colorVec =  phylcol,
+     edgeTranspHigh = 20,
+     title1 = "Network on genus level", 
+     showTitle = TRUE,
+     cexTitle = 2.3,
+     cexNodes = 2,
+     cexLabels = 1,
+     cexHubLabels = 1.3,
+     mar = c(1, 10, 4, 6))
```
If some of these numbers are function defaults, it is not necessary to show the call. This will simplify the example without changing the outcome

antagomir commented on 60_networks.Rmd:
````diff
@@ -0,0 +1,664 @@
+The network plots enable a visual comparison. We use the same configuration as before.
+
+```{r network_plot_diet_difflay, fig.width=17, fig.height=8}
+plot(peerj_netprops_diet,
+     layout = "spring",
+     repulsion = 0.96,
+     rmSingles = TRUE,
+     labelScale = FALSE,
+     nodeSize = "eigenvector",
+     nodeSizeSpread = 3,
+     nodeColor = "cluster", 
+     sameColThresh = 2,
+     edgeTranspHigh = 20,
+     title1 = "Mixed diet", 
+     title2 = "Vegetarian",
+     showTitle = TRUE,
+     cexTitle = 2,
+     cexNodes = 2.5,
+     cexLabels = 1.3,
+     cexHubLabels = 1.5,
+     mar = c(1, 4, 4, 4))
+
````
Same comment for this and subsequent examples.

antagomir commented on 60_networks.Rmd:
```diff
@@ -0,0 +1,664 @@
+All of the above steps are implemented in the R package NetCoMi [@peschel2021netcomi], which provides functionality for the complete workflow of constructing, analyzing, and comparing networks for microbiome data. In this section, we demonstrate the use of NetCoMi.
+
+
+## Intruduction to NetCoMi
```
typo

antagomir commented on 60_networks.Rmd:
```diff
@@ -0,0 +1,664 @@
+
+Since microbial interactions are likely to change between conditions, e.g., between healthy individuals and patients, identifying network differences between groups is often an integral secondary analysis step. Differences can be detected visually by plotting the two networks side by side, or quantitatively using differential network analysis tools.
+
+All of the above steps are implemented in the R package NetCoMi [@peschel2021netcomi], which provides functionality for the complete workflow of constructing, analyzing, and comparing networks for microbiome data. In this section, we demonstrate the use of NetCoMi.
```
I would suggest to formulate this chapter as a chapter on showing how to implement taxonomic network analyses in general, rather than NetCoMi in particular. The contents are good "as is" for now but the chapters should build around the general concept and could allow, in principle, adding comparisons with related alternative approaches with some other package later.

antagomir commented:
Many thanks! General remarks / suggestions:

1. See code comments for suggestions on reducing technicality in the text
2. OMA does not support the phyloseq tool ecosystem, let us discuss separately over email how to solve this. The easiest way could be to add TreeSE support for NetCoMi (instead of converting TreeSE into phyloseq as the chapter currently does). We can provide guidance with this as needed but the material is necessary to update accordingly.
3. Some more references to microbiome network analyses could be added?

antagomir commented:
The automated builds do not pass yet, to troubleshoot.

stefpeschel commented:
Many many thanks for all your comments and suggestions! We have updated the chapter accordingly. It is now more general and includes methods for learning and analyzing microbiome networks that are commonly used in the field.

antagomir commented:
Awesome! 

Could you resolve the conflicts? Let us know if anything unclear on that and support needed.

stefpeschel commented:
Great! I've resolved the conflict but not commited yet. Should I directly commit to master branch or create a new branch for this?

antagomir commented:
You can just commit here on top of the earlier commit.

stefpeschel commented:
Done. Thanks!

antagomir commented on book.bib:
```diff
@@ -1414,6 +1414,181 @@ @Article{ZIBB2018
+@article{ullmann2023over,
+  title = {Over-optimism in unsupervised microbiome analysis: Insights from network learning and clustering},
+  author = {Ullmann, Theresa and Peschel, Stefanie and Finger, Philipp and M{\"u}ller, Christian L and Boulesteix, Anne-Laure},
+  journal = {PLoS Computational Biology},
```
double curly brackets to keep capitals?

antagomir commented on book.bib:
```diff
@@ -1414,6 +1414,181 @@ @Article{ZIBB2018
+@article{faust2012microbial,
+author = {Faust, Karoline and Sathirapongsasuti, J. Fah and Izard, Jacques and Segata, Nicola and Gevers, Dirk and Raes, Jeroen and Huttenhower, Curtis},
+doi = {10.1371/journal.pcbi.1002606},
+journal = {PLoS Computational Biology},
```
same

antagomir commented on book.bib:
```diff
@@ -1414,6 +1414,181 @@ @Article{ZIBB2018
+month = {jul},
+number = {7},
+pages = {e1002606},
+publisher = {Public Library of Science},
```
same

antagomir commented on book.bib:
```diff
@@ -1414,6 +1414,181 @@ @Article{ZIBB2018
+author = {Fang, H and Huang, C and Zhao, H and Deng, M},
+doi = {10.1089/cmb.2017.0054},
+issn = {10665277},
+journal = {Journal of Computational Biology},
```
same

antagomir reviewed (commented):
First comments on bib file

antagomir commented on 60_networks.Rmd:
````diff
@@ -0,0 +1,855 @@
+Note that nodes are sometimes placed too close together so that the labels overlap. You may need to play around with the repulsion argument until you find a value where the labels are legible, but also the clusters are still well recognizable.
+
+```{r network_plot_single_cluster, fig.width=13, fig.height=10}
+plot(spring_netprops,
````
No ggplot equivalent available?

stefpeschel replied:
NetCoMi comes with it's own plot function, which is based on qgraph, which is another state-of-the-art package for network plotting. Considerable effort has been put into ensuring that the function can be used to create visually appealing network plots.

stefpeschel replied:
The idea is also that the plot function expects the object returned by netAnalyze() as input. Thus, it contains all the results of the network analysis, and the user can simply change some arguments in the plot function to highlight different network measures.

antagomir commented on 60_networks.Rmd:
````diff
@@ -0,0 +1,855 @@
+An interesting question to explore now is whether the skin microbiome differs between people whose diets are different.
+
+```{r sample_data}
+table(colData(tse_genus)$Diet)
````
Use altExp?

Also note that colData(tse)$Diet can be written more shortly just tse$Diet

antagomir commented on 60_networks.Rmd:
````diff
@@ -0,0 +1,855 @@
+We split the data set into two phyloseq objects: One with mixed diet subjects and one with vegetarian subjects. Both subsets have nearly the same sample size and are therefore comparable.
+
+```{r split_data}
+tse_mixed <- tse[, colData(tse_genus)$Diet == "Mixed"]
+tse_veg <- tse[, colData(tse_genus)$Diet == "Veg"]
````
How about just:

tse_list <- splitOn(tse, f = "Diet", use_names = TRUE, MARGIN = 2)
tse_list$Veg
tse_list$Mixed

antagomir commented on 60_networks.Rmd:
````diff
@@ -0,0 +1,855 @@
+The `rep.num` argument is set to 10 to perform only 10 repetitions in the model selection approach. This speeds up the permutation tests performed later, and has a negligible effect for this data set.
+
+```{r netConstruct_diet, eval=FALSE}
+spring_net_diet <- netConstruct(data = tse_mixed,
+                                data2 = tse_veg,
````
See above, and use here tse_list$Mixed and tse_list$Veg

antagomir replied:
Or just tse[, tse$Diet=="Mixed"] and tse[, tse$Diet=="Veg"]

antagomir commented on 60_networks.Rmd:
````diff
@@ -0,0 +1,855 @@
+                                seed = 13075)
+```
+
+### Network analysis
+
+All network measures are now computed for both networks. Also, both GCMs are plotted together with a third matrix containing the differences between the GCMs and significance codes that express if the differences are significantly different from zero.
+
+```{r netAnalyze_diet, fig.width=10, fig.height=10}
````
There are many subsections; consider if some of the material could be included in a supplement that is only linked from here, to keep the main text more compact. Optional, for consideration.

muellsen replied:
Thanks for the suggestions! Steffi and I discussed this and we would suggest to split  the chapter into two chapters. The first chapter contains single network estimation and analysis. The second chapter  will be about Differential Network analysis and will comprise 15.6 and 15.7.  The second chapter will then mirror Chapter 11 with DAA and use instead Differential Network Analysis ("DNA"). Like this, the reader may start with the first chapter as intro, and the second chapter assumes, one already knows about network learning...

antagomir replied:
Hi! Apologies for the delays due to travel. This sounds good to me!

antagomir reviewed (changes requested):
Suggestions added, main points include:

- use ggplot and patchwork for plots where possible (consistency with other chapters); alternative ok where this is not possible
- taking more advantage of the capabilities of mia & TreeSE, some suggestions given
- check if the structure can be simplifed a bit further by adding more motivation and intuition at the beginning of each subsection / code, and considering moving some cases to supplementary document that can be in OMA and linked from this page but not in the main chapter / TOC

I hope these help to impove the chapter further before merging. Just request support where necessary. Great work!

antagomir commented:
@ake123 or @TuomasBorman have you bumped into this error, the miaTime pkg seems to fail in GHA although it is listed normally in DESCRIPTION?

stefpeschel commented on 60_networks.Rmd:
```diff
@@ -0,0 +1,855 @@
+library(NetCoMi)
+library(igraph)
+library(mia)
+library(phyloseq)
```
phyloseq is not used anymore. Just forgot to remove the library call.

stefpeschel commented on 60_networks.Rmd:
````diff
@@ -0,0 +1,855 @@
+We start by installing `NetCoMi` from GitHub and loading other necessary packages used in this section.
+
+```{r install-pkg, eval=FALSE}
+if(!require(NetCoMi)){
+  devtools::install_github("stefpeschel/NetCoMi", force = TRUE,
+                         dependencies = c("Depends", "Imports", "LinkingTo"),
+                         repos = c("https://cloud.r-project.org/",
+                                   BiocManager::repositories()))
+}
````
NetCoMi is not yet available on any of these platforms.

antagomir replied:
Doesn't have to be. Good to consider nevertheless. It has many benefits to support broader usage.

stefpeschel commented on 60_networks.Rmd:
````diff
@@ -0,0 +1,855 @@
+dim(tse)
+```
+
+### SparCC correlation
````
In the introduction, we generally refer to association. However, when we talk about a specific method, we refer to the association measure that the method estimates (e.g., SparCC estimates correlation, propr proportionality, and SpiecEasi partial correlation). This should make the user aware of the differences between these association measures.

stefpeschel commented on 60_networks.Rmd:
```diff
@@ -0,0 +1,855 @@
+sparcc_cor03[abs(sparcc_cor03) < 0.3] <- 0
+
+# Compute dissimilarity matrix
+sparcc_diss03 <- sqrt(0.5 * (1 - sparcc_cor03))
```
In fact, this approach is not standard. We propose it in the NetCoMi paper (Peschel et al, 2021) and want to make people aware of the differences in how negative associations are handled. Often, the absolute value of estimated associations is taken, resulting in high edge weights between highly negatively associated nodes. However, from a biological point of view, it also makes sense for negatively associated species to be placed far apart from each other in the network. We have included these two transformations in our workflow to allow users to decide how to handle negative associations. In our opinion, this step is crucial because it affects all distance-based network measures. However, it is often not even mentioned how associations are transformed into edge weights.

The transformations are explained in the introduction, but I will add a comment here as well.

stefpeschel commented on 60_networks.Rmd:
```diff
@@ -0,0 +1,855 @@
+# Turn into graph
+sparcc_graph03 <- adj2igraph(sparcc_sim03)
+
+# Repeat with threshold = 0.4
+sparcc_cor04 <- sparcc_cor
+sparcc_cor04[abs(sparcc_cor04) < 0.4] <- 0
+sparcc_diss04 <- sqrt(0.5 * (1 - sparcc_cor04))
+sparcc_diss04[sparcc_cor04 == 0] <- 1
+sparcc_sim04 <- 1 - sparcc_diss04
+sparcc_graph04 <- adj2igraph(sparcc_sim04)
```
Good point! I will add a function.

stefpeschel commented on 60_networks.Rmd:
````diff
@@ -0,0 +1,855 @@
+prop_est <- as(rho_shrink_est(t(assay(tse, "clr"))), "matrix")
+```
+
+A threshold of 0.4 is used to sparsify the estimated proportionality matrix.
+
````
All the methods used in the examples are explained in the introduction of this chapter below the flowchart. We thought it would be better to have a section at the beginning that introduces the methods and puts them in context, so that the examples don't need further explanation. In the introduction, we could add cross-links to the examples so it's clearer which example belongs to which method. Or would you prefer to have the explanations above each application?

stefpeschel commented on 60_networks.Rmd:
````diff
@@ -0,0 +1,855 @@
+set.seed(13075)
+lay_fr <- layout_with_fr(se_mb_graph)
+
+par(mfrow = c(2,2))
+plot(sparcc_graph03, layout = lay_fr, vertex.size = vsize, 
+     vertex.label = NA, main = "SparCC (thresh 0.3)")
+plot(sparcc_graph04, layout = lay_fr, vertex.size = vsize, 
+     vertex.label = NA, main = "SparCC (thresh 0.4)")
+plot(prop_graph, layout = lay_fr, vertex.size = vsize, 
+     vertex.label = NA, main = "Shrinkage proportionality\n(thresh 0.4)")
+plot(se_mb_graph, layout = lay_fr, vertex.size = vsize, 
+     vertex.label = NA, main = "SpiecEasi (MB)")
+```
````
The igraph package (used here) is a state-of-the-art package for visualizing networks in R. We could use the [ggnet2](https://briatte.github.io/ggnet/) package instead but I don't see any advantage because the output would look quite the same.

stefpeschel commented on 60_networks.Rmd:
````diff
@@ -0,0 +1,855 @@
+```{r network_plot_single_phylum, fig.width=15, fig.height=10}
+# Get phyla names
+taxtab <- as.data.frame(rowData(tse_genus))
+phyla <- gsub("p__", "", taxtab[, "phylum"])
````
Even though it's not a general preprocessing step? The phyla vector is needed only here for the plot. One would not need this step if another coloring is used for the plot.

antagomir commented:
Just let me know when the iteration is ready for a new review, so I will do it all at once.

ake123 commented:
> @ake123 or @TuomasBorman have you bumped into this error, the miaTime pkg seems to fail in GHA although it is listed normally in DESCRIPTION?

Yes, my last PR had the same issue

antagomir commented:
> > @ake123 or @TuomasBorman have you bumped into this error, the miaTime pkg seems to fail in GHA although it is listed normally in DESCRIPTION?
> 
> Yes, my last PR had the same issue

Ok great then this should become solved as soon as we merge your PR and sync this one.

ake123 commented:
I have fixed it and now making similar corrections with `ggord`

antagomir commented:
> I believe the package must be in CRAN or Bioc. Otherwise it cannot be listed in DESCRIPTION

Hmm but the DESCRIPTION contains many packages like

Remotes:
    github::microbiome/microbiomeDataSets,
    github::microbiome/mia,
    github::microbiome/miaViz,
    github::microbiome/miaSim,        
    github::microbiome/miaTime,
    github::HenrikBengtsson/matrixStats,
    github::biobakery/Maaslin2,
    github::SchlossLab/mikropml,
    github::jsilve24/fido

stefpeschel commented:
We have updated the network chapter according to your suggestions. Thanks again for your detailed review! 
As Christian has already announced, we have split the chapter into two separate chapters: "Network Learning and Analysis" and "Network Comparison".  The chapters are now ready to be reviewed again.

antagomir commented on 60_network_learning.Rmd:
```diff
@@ -0,0 +1,654 @@
+        -   CCREPE ([ccrepe](https://bioconductor.org/packages/release/bioc/html/ccrepe.html) package)
+        -   CCLasso ([R code on GitHub](https://github.com/huayingfang/CCLasso))
+
+    -   **Conditional dependence:** Since standard correlations include both direct and indirect dependencies, conditional dependence or partial correlation is often preferred for measuring association. Unlike (marginal) correlation, it expresses the relationship between two features conditioned on all other features in the data set. The approach and R package [SpiecEasi](https://github.com/zdk12 ...
```
aforementionedconditional -> space?

antagomir commented on 60_network_learning.Rmd:
```diff
@@ -0,0 +1,654 @@
+# Network learning and analysis {#network-learning}
+
```
The intro chapter contains a lot of useful information. 

I am however thinking if we should get the user busy with recommended analysis faster, and have for instance in the end of this chapter a subsection that provides more details. Then the intro could be shortened, and it could cite that concluding section? 

I am mainly thinking about a user who just wants to get started with trying out some recommended approach, and then maybe check more methods and details after gaining some experience with one recommended method as a starting point?

stefpeschel replied:
So, I've completely reorganized the chapter. The first sections now only show network learning with the SPRING method. And I added a section at the end, where the three types of association are compared.

antagomir commented on 60_network_learning.Rmd:
````diff
@@ -0,0 +1,654 @@
+
+```{r preprocessing}
+# Agglomerate by genus and subset by prevalence
+tse <- subsetByPrevalentFeatures(tse,
+                                 rank = "genus",
+                                 prevalence = 0.2,
+                                 detection = 0)
````
This does not agglomerate (although the text above says so), this filters out the rare. 

Agglomeration would instead happen with:

`tse <- mergeFeaturesByPrevalence(tse, rank = "genus", prevalence = 0.2, detection = 0)`

antagomir commented on 60_network_learning.Rmd:
````diff
@@ -0,0 +1,654 @@
+Since we want to use the phylum names for node colors later in the network plot, we assign low abundant phyla to a new class "Other" to reduce the number of colors needed in the plot.
+
+```{r}
+# Assign low abundant phyla to a "Other" class
+phyla <- rowData(tse)$phylum
+low_abundant <- names(table(phyla)[table(phyla) <= 3])
+phyla[phyla %in% low_abundant] <- "Other"
+rowData(tse)$phylum <- phyla
+
+# Take a look at the table
+table(rowData(tse)$phylum)
````
I suggest two small changes for clarity. Ok?

1. Now "rare" phyla are detected based on how many *rows* they have; this counts the number of different ASVs (or other features) within the phylum, and not their abundance although the text - in my view misleadingly - says that these would be "low abundant phyla"; this is not necessarily the case. I think that abundance/prevalence-based agglomeration would be better justified? For that you could do 

```
# Detect rare features (this is complement of prevalent)
rare.taxa <- getRareFeatures(tse, rank="Phylum", detection=0, prevalence=20/100)

# Merge non-prevalent features into "Other" group
rowData(tse)$Phylum_merged <- stringi::stri_replace_all_regex(rowData(tse)$Phylum, pattern=rare.taxa, replacement="Other", vectorize=FALSE)

# There may also be NA phyla - decide what to do with those; perhaps just discard those rows where phylum is not identified?
```



2. Adding a new column, for instance phylum_merged, instead of overwriting original phyla. Then both will remain available.

antagomir commented on 60_network_learning.Rmd:
````diff
@@ -0,0 +1,654 @@
+```
+
+
+### SparCC correlation {#sparcc-correlation}
````
Because nowadays SPIEC-EASI and other methods are usually replacing SparCC I am wondering if SparCC and some other methods are really needed for the main text. I would separate the material in two categories: 1) recommended practice & 2) educational background. If SparCC is not recommended for most real scenarios (?) then could we think of creating a supplementary document that can be located in OMA and cited in the main text, such document could have examples of methods that we would not recommend for most of the standard use cases but that may still be useful as educational background? 

But we may have different information & thoughts, open to discuss the choices.

antagomir commented on 60_network_learning.Rmd:
````diff
@@ -0,0 +1,654 @@
+
+```{r nets_sparcc_prop_se, fig.height=10, fig.width=10}
+# Node sizes
+vsize <- colMeans(t(assay(tse, "clr"))) + 6
````
I am not sure if mean over CLR is meaningful measure for node size? CLR can include both neg and pos values?

antagomir commented on 60_network_learning.Rmd:
````diff
@@ -0,0 +1,654 @@
+```{r select_nodes}
+# Function for selecting taxa with highest sum of edge weights
+select_taxa <- function(adja, ntaxa = 50) {
+  sel <- names(sort(rowSums(adja), decreasing = TRUE))[1:ntaxa]
````
use `seq_len(ntaxa)` instead of `1:ntaxa` (current Bioc requirement & best practice)

antagomir commented on 60_network_learning.Rmd:
````diff
@@ -0,0 +1,654 @@
+
+```{r clustered_heatmap, fig.width=10, fig.height=10}
+# Color vector for the legend
+col <- colorRamp2(c(-1, -0.5, 0, 0.5, 1), 
+                  c("royalblue4", "lightblue", "white", "orange", "firebrick3"))
````
Consider using the microshades R package accessible palettes:
https://karstenslab.github.io/microshades/

-> this is designed for microbiome data and seems like nice work to support

stefpeschel replied:
This package seems to be quite neat and I tried to use it here. Unfortunately, this would make the example too complicated, because microshades needs phyloseq as an input, so we would have to transform the tse object first. Also, microshades creates a gglot object, which cannot be used for the network plot. So I would use a color palette from RColorBrewer instead.

antagomir replied:
ok great

antagomir commented on 61_network_comparison.Rmd:
```diff
@@ -0,0 +1,323 @@
+# Network comparison {#network-comparison}
```
same comment than in the first chapter; consider if some educational background could go to the end of the chapter as extra info; or into a separate supplementary doc that is not in table of contents but can be in OMA files and linked from here.

stefpeschel replied:
I moved parts of the introduction to the end of the chapter.

antagomir commented on 61_network_comparison.Rmd:
````diff
@@ -0,0 +1,323 @@
+
+```{r preprocessing}
+# Agglomerate by genus and subset by prevalence
+tse <- subsetByPrevalentFeatures(tse,
+                                 rank = "genus",
+                                 prevalence = 0.2,
+                                 detection = 0)
+
````
Note comments from the earlier chapter. This is not agglomerating, this is filtering.

antagomir commented on 61_network_comparison.Rmd:
```diff
@@ -0,0 +1,323 @@
+tse <- transformAssay(tse, method="clr", pseudocount=1)
+
+# Assign low abundant phyla to a "Other" class
+phyla <- rowData(tse)$phylum
+low_abundant <- names(table(phyla)[table(phyla) <= 3])
+phyla[phyla %in% low_abundant] <- "Other"
+rowData(tse)$phylum <- phyla
```
See suggestion in the previous chapter.

antagomir reviewed (changes requested):
Very nice! Just a few remarks/suggestions.

antagomir commented:
Hi @stefpeschel - just confirm - you are still working on this?

stefpeschel commented:
Yes, I am still working on it. Since I have to rewrite all the examples where I compare all four association measures, it is again a bit of work to move parts of the chapter to other sections. I will upload the new version this week.

antagomir commented:
Awesome - thinking about future maintenance load is also good consideration indeed. No hurry from our side, just checking.

stefpeschel commented:
I followed your suggestions and reorganized the chapter. The new version is pushed.

antagomir commented:
Thanks!
